File: RB/cpp/RiotBase/System/PatchId.cpp
#include "PatchId.h"
#include "UsrId.h"

namespace RiotBase {

void PatchId::write_to_stream( std::ostream &os ) const {
    os << num << "@" << ( num ? dev : DevId{} );
}

void glo_to_loc( PatchId &pid, DevId cur_dev, UsrId cur_usr) {
    if ( pid.num and pid.dev == cur_dev ) pid.dev.val = 0;
}

void loc_to_glo( PatchId &pid, DevId cur_dev, UsrId cur_usr ) {
    if ( pid.num and pid.dev.val == 0 ) pid.dev = cur_dev;
}

} // namespace RiotBase

File: RB/ts/RiotBase/System/PatchId.ts
import DevId from "./DevId";

//
export default
class PatchId {
    constructor( dev : DevId = new DevId, num : number = 0 ) {
        this.dev = dev;
        this.num = num;
    }

    toString( cur_dev? : DevId ) : string {
        return `${ this.num }@${ this.dev.toString( cur_dev ) }`
    }

    static fromString( str : string ) : PatchId {
        let i = str.indexOf( "@" );
        return new PatchId( new DevId( str.substr( i + 1 ) ), parseInt( str.substr( 0, i ) ) );
    }

    self_glo_to_loc( dev : DevId ) : void {
        this.dev.self_glo_to_loc( dev );
    }

    self_loc_to_glo( dev : DevId ) : void {
        this.dev.self_loc_to_glo( dev );
    }

    inf( p : PatchId ) : boolean {
        return this.num == p.num ? this.dev.inf( p.dev ) : this.num < p.num;
    }

    sup( p : PatchId ) : boolean {
        return this.num == p.num ? this.dev.sup( p.dev ) : this.num > p.num;
    }

    equ( p : PatchId ) : boolean {
        return this.num == p.num && this.dev.equ( p.dev );
    }

    dev: DevId;
    num: number;
}
